tools/smoother_benchmarking/metrics.py

```diff
@@ -0,0 +1,173 @@
+            quad = euler2quat(0.0, 0.0, yaw)
+            start.pose.orientation.w = quad[0]
+            start.pose.orientation.x = quad[1]
+            start.pose.orientation.y = quad[2]
+            start.pose.orientation.z = quad[3]
+            break
+    return start
+
+def getRandomGoal(costmap, start, max_cost, side_buffer, time_stamp, res):
+    goal = PoseStamped()
+    goal.header.frame_id = 'map'
+    goal.header.stamp = time_stamp
+    while True:
+        row = randint(side_buffer, costmap.shape[0]-side_buffer)
+        col = randint(side_buffer, costmap.shape[1]-side_buffer)
+
+        start_x = start.pose.position.x
+        start_y = start.pose.position.y
+        goal_x = col*res
+        goal_y = row*res
+        x_diff = goal_x - start_x
+        y_diff = goal_y - start_y
+        dist = math.sqrt(x_diff ** 2 + y_diff ** 2)
+
+        if costmap[row, col] < max_cost and dist > 3.0:
+            goal.pose.position.x = goal_x
+            goal.pose.position.y = goal_y
+
+            yaw = uniform(0, 1) * 2*math.pi
+            quad = euler2quat(0.0, 0.0, yaw)
+            goal.pose.orientation.w = quad[0]
+            goal.pose.orientation.x = quad[1]
+            goal.pose.orientation.y = quad[2]
+            goal.pose.orientation.z = quad[3]
+            break
+    return goal
+
+def main():
+    rclpy.init()
+
+    navigator = BasicNavigator()
+
+    # Set our experiment's initial pose
+    initial_pose = PoseStamped()
+    initial_pose.header.frame_id = 'map'
+    initial_pose.header.stamp = navigator.get_clock().now().to_msg()
+    initial_pose.pose.position.x = 1.0
+    initial_pose.pose.position.y = 1.0
+    initial_pose.pose.orientation.z = 0.0
+    initial_pose.pose.orientation.w = 1.0
+    navigator.setInitialPose(initial_pose)
+
+    # Wait for navigation to fully activate
+    print("Waiting for nav2 to activate")
+    navigator.waitUntilNav2Active()
+
+    # Get the costmap for start/goal validation
+    costmap_msg = navigator.getGlobalCostmap()
+    costmap = np.asarray(costmap_msg.data)
+    costmap.resize(costmap_msg.metadata.size_y, costmap_msg.metadata.size_x)
+
+    # planner: 'NavFn', 'SmacLattice', 'SmacHybrid', 'ThetaStarPlanner'
+    planner = 'SmacHybrid'
+    smoothers = ['simple_smoother', 'constrained_smoother']
+    max_cost = 210
+    side_buffer = 10
+    time_stamp = navigator.get_clock().now().to_msg()
+    results = []
+    seed(33)
+
+    random_pairs = 100
+    i = 0
+    res = costmap_msg.metadata.resolution
+    while i < random_pairs:
+        print("Cycle: ", i, "out of: ", random_pairs)
+        start = getRandomStart(costmap, max_cost, side_buffer, time_stamp, res)
+        goal = getRandomGoal(costmap, start, max_cost, side_buffer, time_stamp, res)
+        print("Start", start)
+        print("Goal", goal)
+        result = getPlannerResults(navigator, start, goal, planner)
+        if result is not None:
+            print("MAO: result = ", result)
```

Comment:
MAO?

Reply:
That was the initials of my second, first and middle name, I've used for debugging :smiley: Removed in the latest update.

Reply:
Ok! I use "STEVE" for similar so I know what you mean.  Just something easy your brain can pick up the pattern in easily in a dense log file